Improve performance of DataObject magic calls and addData()

igorwulff opened the pull request:
The magic calls from the DataObject are being called over a 100k when there is no block_cache available and depending on the page often thousands of times when all block_cache is available.

The small change for addData() is just included because its a tiny optimization and prevents a bug from happening. When a $value is an array the original call to setData would override the internal $_data array, this is not expected behavior and just setting it directly to $_data is faster anyway.

The other change is in the __call magic method and the _underscore function and the change is quite radical. Since this method can be called so often, we want to prevent further function calls if possible. Functional calls are normally not that expensive, but will become more expensive if you use this extensively (thousands of even hundred of thousands extra calls):

These are my own benchmarks based on a local environment with opcache and with PROD mode. 

HOMEPAGE transaction
- All cache enabled except for FPC (__call is called 975 times): 4.318307ms -> 1.469747ms 
- All cache enabled except for FPC + Block_html  (__call is called 100379 times): 385.36412ms -> 88.412914ms

PLP transaction:
- All cache enabled except for FPC  (__call is called 4628 times): 19.625463ms -> 5.754514ms 
- All cache enabled except for FPC + Block_html (__call is called 112591 times): 404.84066ms -> 99.390374ms

Now this PR also skips some checks. It assumes if a $method call start with 'g' it will be for 'get' for example, this is done in my PR to get the most optimal performance result but might not be preferred. I also removed an index check with $args[1] for the setter since it isn't used anywhere, but might also be something we would want to change.

<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.
-->

### Related Pull Requests
<!-- related pull request placeholder -->

### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
1. Fixes magento/magento2#<issue_number>

### Manual testing scenarios (*)
<!---
    Please provide a set of unambiguous steps to test the proposed code change.
    Giving us manual testing scenarios will help with the processing and validation process.
-->
1. ...
2. ...

### Questions or comments
<!---
	If relevant, here you can ask questions or provide comments on your pull request for the reviewer
	For example if you need assistance with writing tests or would like some feedback on one of your development ideas
-->

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)


### Resolved issues:
1. [x] resolves magento/magento2#37637: Improve performance of DataObject magic calls and addData()

igorwulff committed:
Improve performance of DataObject magic calls and addData()

The magic calls from the DataObject are being called over a 100k when there is no block_cache available and depending on the page often thousands of times when all block_cache is available.

The small change for addData() is just included because its a tiny optimization and prevents a bug from happening. When a $value is an array the original call to setData would override the internal $_data array, this is not expected behavior and just setting it directly to $_data is faster anyway.

The other change is in the __call magic method and the _underscore function and the change is quite radical.
Since this method can be called so often, we want to prevent further function calls if possible. Functional calls are normally not that expensive, but will become more expensive if you use this extensively (thousands of even hundred of thousands extra calls):

These are my own benchmarks based on a local environment with opcache and with PROD mode. 

HOMEPAGE transaction
All cache enabled except for FPC (__call is called 975 times): 4.318307ms -> 1.469747ms
All cache enabled except for FPC + Block_html  (__call is called 100379 times): 385.36412ms -> 88.412914ms

PLP transaction:
All cache enabled except for FPC  (__call is called 4628 times): 19.625463ms -> 5.754514ms
All cache enabled except for FPC + Block_html (__call is called 112591 times): 404.84066ms -> 99.390374ms

Now this PR also skips some checks. It assumes if a $method call start with 'g' it will be for 'get' for example, this is done in my PR to get the most optimal performance result but might not be preferred. I also removed an index check with $args[1] for the setter since it isn't used anywhere, but might also be something we would want to change.

igorwulff committed:
Update DataObject.php

ihor-sviziev commented on lib/internal/Magento/Framework/DataObject.php:
```diff
@@ -387,22 +387,26 @@ public function toString($format = '')
-                $key = $this->_underscore(substr($method, 3));
-                return isset($this->_data[$key]);
+        switch($method[0]) {
+            case 'g':
```
I don't think we can check the only first letter. 
Examples: `$dataObject->greatMethod('test')` earlier were throwing exception, but now will go into this case.

igorwulff replied:
I tested and tried varied things, and reverted it back for now... performance gains are a bit less, but still very good:
Homepage without block_cache is now: 158.512655
PLP without block_cache is now: 172.213267

igorwulff committed:
Reverted back the way magic __call method types (set, get, etc...) are identified

ihor-sviziev commented:
@magento run all tests

igorwulff commented:
Hi @ihor-sviziev We've been analyzing and testing this specific patch and think a change is required for the magic 'get' call, since we are seeing some issues in various areas, what do you think?

```
case 'get':
    if (isset($args[0]) && $args[0] !== null) {
        return $this->getData(
            self::$_underscoreCache[$method] ?? $this->_underscore($method),
                $args[0]
        );
    }

    return $this->_data[
        self::$_underscoreCache[$method] ?? $this->_underscore($method)
    ] ?? null;
```

igorwulff committed:
Fix issues with magic __call 'get' calls in certain conditions

Fix issues with magic __call 'get' calls in certain conditions

ihor-sviziev committed:
Merge branch '2.4-develop' into patch-6

ihor-sviziev commented:
@magento run all tests

igorwulff committed:
Regression fixes to __call method dataobject

Added fix for regression issues due to changes in __call method in dataobject for setter calls.
Added small optimization tweak to skip substr calls in switch statement

igorwulff committed:
Merge branch '2.4-develop' into patch-6

igorwulff committed:
Update DataObjectTest.php

igorwulff committed:
Fix for missing + sign in preg_replace in _underscore call

igorwulff commented:
@magento run all tests

igorwulff committed:
phpstan issues for dataobject.php changes

igorwulff commented:
@magento run Static Tests

igorwulff committed:
Further changes to fix unit tests and keep things backwards compatible.

I'm really not happy with how Magento is using the magic __call method and I would propose to deprecated the behavior.

igorwulff commented:
@magento run Unit Tests

igorwulff committed:
Regression fix from unit tests for __call method

igorwulff commented:
@magento run Unit Tests

igorwulff committed:
Merge branch '2.4-develop' into patch-6

igorwulff commented:
@magento run all tests

hostep commented on lib/internal/Magento/Framework/DataObject.php:
```diff
@@ -387,22 +387,38 @@ public function toString($format = '')
     public function __call($method, $args)
     {
-        switch (substr((string)$method, 0, 3)) {
+        switch ($method[0].($method[1] ?? '').($method[2] ?? '')) {
```
Can we put this `$method[0].($method[1] ?? '').($method[2] ?? '')` into a variable? Not sure if this will have an extra performance impact but it should at least make the code a bit more readable 😄

ihor-sviziev replied:
I added a commend to make it more obvious and not affect performance improvement

igorwulff committed:
Fix setData function calls due to dependencies

setData is expected to be called from various dependencies where addData is being used, so the original change has been reverted.

ihor-sviziev committed:
Merge branch '2.4-develop' into patch-6

ihor-sviziev commented:
@magento run all tests

ihor-sviziev committed:
Improve performance of DataObject magic calls and addData

Fix failing static tests

ihor-sviziev committed:
Improve performance of DataObject magic calls and addData

Add comment to not obvious switch condition

ihor-sviziev committed:
Merge branch '2.4-develop' into patch-6

ihor-sviziev commented:
@magento run all tests

ihor-sviziev reviewed (changes requested):
Hi @igorwulff,
Looks like your changes broke the integration tests. Could you please double-check that?

igorwulff commented:
> Hi @igorwulff, Looks like your changes broke the integration tests. Could you please double-check that?

I will look into it. 

I also saw an issue with vendor/magento/module-paypal/Model/Payflow/Request.php where the DataObject is extended. The logic in the overridden __call method does some weird things with changing the key when a = or & char is seen in the $args variable:
```
        if (isset($args[0]) && (strstr($args[0], '=') || strstr($args[0], '&'))) {
            $key .= '[' . strlen($args[0]) . ']';
        }
```

I don't understand why this is used like this.

It should be fixed by just changing the line:
```
$key = $this->_underscore(substr($method, 3));
```

To
```
$key = $this->_underscore($method);
```

ihor-sviziev commented:
Hi @igorwulff,
Any updates on this? :)

engcom-Hotel committed:
Merge branch '2.4-develop' into patch-6

engcom-Hotel commented:
@magento run all tests

engcom-Hotel committed:
Merge branch '2.4-develop' into patch-6

glo24157 committed:
Fix Integration test failure for testSuccessfulNotification test

engcom-Hotel commented:
@magento run all tests

engcom-Hotel commented:
@magento run all tests

igorwulff committed:
Add enter in docblock to fix phpcs error

Fix:
PHP Code Sniffer detected 1 violation(s): 

FILE: /var/www/html/app/code/Magento/Paypal/Model/Payflow/Request.php
-----------------------------------------------------------------------------------------------
FOUND 0 ERRORS AND 1 WARNING AFFECTING 1 LINE
-----------------------------------------------------------------------------------------------
 17 | WARNING | There must be exactly one blank line between lines short and long descriptions
-----------------------------------------------------------------------------------------------

igorwulff commented:
@magento run all tests

engcom-Hotel commented:
@magento run Integration Tests

glo24157 committed:
Resolve B2B integration tests failures

engcom-Hotel commented:
@magento run all tests

engcom-Hotel commented:
@magento run WebAPI Tests

glo24157 committed:
Fix Static tests failures

engcom-Hotel commented:
Fixed static test failures. Functional and WebApi tests failures seem flaky to me.


@magento run all tests

engcom-Hotel commented:
@magento run WebAPI Tests, Functional Tests CE

engcom-Hotel commented:
@magento run Functional Tests CE, Functional Tests EE, Functional Tests B2B

engcom-Lima commented:
@magento run Functional Tests B2B , Functional Tests EE, Functional Tests CE

engcom-Lima commented:
@magento create issue

engcom-Echo commented:
@engcom-Hotel moving this to `Pending Review` code approval is still pending for it.

ihor-sviziev reviewed (changes requested):
@igorwulff, could you please double-check the performance impact after all changes were done recently to fix test failures?

igorwulff commented:
Hi @ihor-sviziev the changes were minimal and mostly esthetic, so this did not impact the performance improvements. So I think we are good to go to proceed.

igorwulff committed:
Merge branch '2.4-develop' into patch-6

engcom-Hotel commented:
Hello @igorwulff,

Thanks for the response!

As per your comment, we can move further with this PR, hence moving this PR to `Pending Review`.

Thanks

hostep commented:
> @engcom-Hotel engcom-Hotel requested a review from hostep 20 hours ago

Am I supposed to do anything here? Kind of annoying that once you leave one little remark, you're supposed to do a full review 😝

ihor-sviziev commented:
@magento run all tests

ihor-sviziev commented:
@magento run Functional Tests B2B, Functional Tests EE, WebAPI Tests

engcom-Hotel committed:
Merge branch '2.4-develop' into patch-6

engcom-Hotel commented:
@magento run all tests

engcom-Hotel commented:
@magento run all tests

engcom-Hotel commented:
Hello @igorwulff,

While testing this PR we have below observations. 

In order to test the performance improvement we have used the below profiler:

https://github.com/NoiseByNorthwest/php-spx

### HomePage: With FPC Disable

_**2.4-develop**_
![2 4-develop_homepage](https://github.com/magento/magento2/assets/51681618/1b042c3e-d8ae-4868-b0e1-1b721cec42ea)

**_After PR Changes_**
![PR_homepage](https://github.com/magento/magento2/assets/51681618/2369723a-03b1-4272-9eba-eea8dcdc49bc)

### HomePage: With FPC and block_html cache Disable

_**2.4-develop**_
![2 4-develop_homepage_fpc_blockhtml](https://github.com/magento/magento2/assets/51681618/9f633a7c-53ba-4792-9937-22730c198ade)


**_After PR Changes_**
![PR_homepage_fpc_blockhtml](https://github.com/magento/magento2/assets/51681618/ee9a30ce-8093-4ae8-83c9-1f4c9f88d5b8)

### PLP: With FPC Disable

**_2.4-develop_**
![2 4-develop_PLP](https://github.com/magento/magento2/assets/51681618/94d927b1-4fc0-430c-9666-414d0db3c350)

_**After PR Changes**_
![PR_PLP](https://github.com/magento/magento2/assets/51681618/32c8e456-4b0a-4ac8-b175-394dc1a093c5)

### PLP: With FPC and block_html cache Disable

**_2.4-develop_**
![2 4-develop_PLP_fpc_blockhtml](https://github.com/magento/magento2/assets/51681618/8eb43897-1a6e-4bd3-9d5a-003f23512e26)

_**After PR Changes**_
![PR_PLP_fpc_blockhtml](https://github.com/magento/magento2/assets/51681618/74062517-025c-4d8c-b625-aca576482645)


As per the above screenshot, we can see the major improvements in the HomePage rendering after PR changes but it looks similar in the case of PLP after and before PR changes. 

We request you to please look into it. Also please let us know the tool used by you for performance improvement.

Thanks

engcom-Hotel commented:
Hello @igorwulff,

We have tested this PR, and we can see performance improvements on the home page, but it seems there is not much performance improvement visible to us for PLP. 

Please go through the detailed comment here https://github.com/magento/magento2/pull/36345#issuecomment-1634060205.

Please let us know which tool you have used for performance measurements. Till the time we are moving this PR `On Hold`.

Thanks

engcom-Hotel commented:
Hello @igorwulff,

Thank you for your contribution!

As per the comment https://github.com/magento/magento2/pull/36345#issuecomment-1634060205, it seems that the PLP page needs more performance improvement. Hence we are closing this PR for now.

Please feel free to reopen or let us know when you are ready to work on it again, and we will be happy to re-open it.

Thank you!

hostep commented:
So you are closing a proven solution just because it doesn't match exactly with the numbers in the description?

@ihor-sviziev what do you make of this? If it improves performance, even if not 100% exact the same as written in the description and causes no regressions, why don't we go through with it?

engcom-Hotel committed:
Merge branch '2.4-develop' into patch-6

engcom-Hotel commented:
@magento run all tests

engcom-Hotel commented:
@magento run all tests

engcom-Hotel commented:
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE, Integration Tests, Unit Tests, WebAPI Tests

engcom-Hotel commented:
As per this comment https://github.com/magento/magento2/pull/36345#issuecomment-1634060205, there are some improvements visible in this PR for HomePage and PDP. 

Hence we can move forward with the PR, but due to some tests are still failing, moving this PR into `Extended Testing`.

Thanks

engcom-Hotel commented:
@magento run Functional Tests CE, Functional Tests EE, Unit Tests, WebAPI Tests

engcom-Hotel commented:
@magento run Functional Tests EE, WebAPI Tests

engcom-Hotel commented:
@magento run Functional Tests EE

engcom-Hotel commented:
All tests are green, hence moving this PR into `Merge in Progress`
